src/linear_algebra/invariant_submodule.lean

```diff
@@ -0,0 +1,125 @@
+lemma invariant_under_iff_map (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ U.map T ≤ U := submodule.map_le_iff_le_comap.symm
+
+/-- `U` is `T` invariant if and only if `set.maps_to T U U` -/
+lemma invariant_under_iff_maps_to (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ set.maps_to T U U := iff.rfl
+
+/-- `U` is `T` invariant is equivalent to saying `T(U) ⊆ U` -/
+lemma invariant_under_iff (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ T '' U ⊆ U := by rw [← set.maps_to', U.invariant_under_iff_maps_to]
+
+variables (U V : submodule R E) (hUV : is_compl U V) (T : E →ₗ[R] E)
+
+local notation `pᵤ` := submodule.linear_proj_of_is_compl U V hUV
+local notation `pᵥ` := submodule.linear_proj_of_is_compl V U hUV.symm
+
+lemma proj_comp_self_comp_proj_eq_of_invariant_under
+  (h : U.invariant_under T) (x : E) : ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x) :=
+begin
+  rw submodule.linear_proj_of_is_compl_eq_self_iff,
+  exact h (submodule.coe_mem _),
+end
+
+lemma invariant_under_of_proj_comp_self_comp_proj_eq
+  (h : ∀ x : E, ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x)) : U.invariant_under T :=
+begin
+  intros u hu,
+  rw [submodule.mem_comap, ← submodule.linear_proj_of_is_compl_eq_self_iff hUV _,
+      ← (submodule.linear_proj_of_is_compl_eq_self_iff hUV u).mpr hu, h],
+end
+
+lemma proj_comp_self_comp_proj_eq_iff_invariant_under :
+  U.invariant_under T ↔ (∀ x : E, ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x)) :=
+⟨U.proj_comp_self_comp_proj_eq_of_invariant_under V hUV T,
+ U.invariant_under_of_proj_comp_self_comp_proj_eq V hUV T⟩
+
+lemma proj_comp_self_comp_proj_eq_iff_invariant_under' :
+  V.invariant_under T ↔ (∀ x : E, (pᵤ (T ↑(pᵤ x)) : E) = pᵤ (T x)) :=
+by simp_rw [submodule.proj_comp_self_comp_proj_eq_iff_invariant_under _ _ hUV.symm,
+            linear_proj_of_is_compl_eq_self_sub_linear_proj, map_sub, sub_eq_self,
+            submodule.coe_sub, sub_eq_zero, eq_comm]
+
+lemma compl_invariant_under_iff_linear_proj_and_T_commute :
+  (U.invariant_under T ∧ V.invariant_under T) ↔ commute (U.subtype.comp (pᵤ)) T :=
+begin
+  simp_rw [commute, semiconj_by, linear_map.ext_iff, linear_map.mul_apply,
+           linear_map.comp_apply, U.subtype_apply],
+  split,
+  { rintros ⟨h1, h2⟩ x,
+    rw [← (U.proj_comp_self_comp_proj_eq_iff_invariant_under' V _ _).mp h2 x],
+    exact (submodule.linear_proj_of_is_compl_eq_self_iff hUV _).mpr
+      (h1 (set_like.coe_mem (pᵤ x))) },
+  { intros h,
+    split,
+    { simp_rw [U.proj_comp_self_comp_proj_eq_iff_invariant_under _ hUV, h,
+               submodule.linear_proj_of_is_compl_idempotent hUV],
+      exact λ x, rfl },
+    { simp_rw [U.proj_comp_self_comp_proj_eq_iff_invariant_under' _ hUV, h,
+               submodule.linear_proj_of_is_compl_idempotent hUV],
+      exact λ x, rfl } }
+end
+
+lemma commutes_with_linear_proj_iff_linear_proj_eq [invertible T] :
+  commute (U.subtype.comp pᵤ) T ↔
+    (⅟ T).comp ((U.subtype.comp pᵤ).comp T) = U.subtype.comp pᵤ :=
```

Comment:
This can be generalized to any linear_map in place of `U.subtype.comp pᵤ`, right?

Reply:
Yeah! Probably better to rename it to something like `commutes_with_invertible_linear_map_iff_eq_self`, right?

Should I keep it in this file or should I put it in `linear_algebra/basic`?

Reply:
Probably due to [function.commute](https://leanprover-community.github.io/mathlib_docs/logic/function/conjugate.html#function.commute)? You can write `_root_.commute` instead.

Reply:
done